libs/jwt-tools/src/Jwt/Tools.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE ForeignFunctionInterface #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2022 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Jwt.Tools (generateDpopToken) where
+
+import Control.Exception
+import Control.Monad.Trans.Except
+import Data.ByteString.Conversion (ToByteString, toByteString')
+import Data.Id (ClientId (client))
+import Data.Misc (HttpsUrl)
+import Data.Nonce (Nonce)
+import Data.PEMKeys
+import Data.String.Conversions (cs)
+import Foreign.C (CUChar (..))
+import Foreign.C.String (CString, newCString, peekCString)
+import Foreign.Ptr (Ptr, nullPtr)
+import Foreign.Storable (peek)
+import Imports
+import Network.HTTP.Types (StdMethod (..))
+import Numeric (readHex)
+import Wire.API.MLS.Credential
+import Wire.API.MLS.Epoch (Epoch (..))
+import Wire.API.User.Client.DPoPAccessToken
+
+data JwtResponse
+
+foreign import ccall unsafe "generate_dpop_access_token"
+  generate_dpop_access_token ::
+    CString ->
```

Comment:
I wonder if it's worth adding comments or type aliases to these arguments for legibility?